### ISSUE TITLE
Custom objective function

### ISSUE THREAD

**Issue by link-er** [fchollet/keras#3561]
Hello!
I am trying to define custom objective function for my network and I ran into several questions, would be very grateful if you can help me to solve them!
1. As I see in the source of objectives (https://github.com/fchollet/keras/blob/master/keras/objectives.py#L48) the error will be taken as mean value **along last axis**. I have network that have as input 3D volume and as the output again 3D volume - I am making a prediction for each voxel. So the dimensionality of the arrays in the objective is (5,1,215,815,215) (I have only 5 training volumes). And I really do not understand whether it is even valid to take the error as mean on the last dimension. 
2. Are there any native methods for using weighted objective? I have highly unbalanced distribution of classes inside the volume and I need to add weights to get correct classification after training.

Thanks a lot!

**Comment by carlthome** [comment]
1. Could you reshape your data as a single vector perhaps? E.g. this old trick
```C
int array[width * height];
array[width * row + col] = value;  
```
The mean will be the mean of all dimensions of course, but that must be done anyway because the optimizer expects a scalar loss.

You could consider having multiple inputs/outputs (see [the Functional API](https://keras.io/getting-started/functional-api-guide/#multi-input-and-multi-output-models)) for the three dimensions, but a reshape would probably be easier.

2. Yes. [class weights](https://github.com/fchollet/keras/blob/master/keras/models.py#L580), [sample weights](https://github.com/fchollet/keras/blob/master/keras/models.py#L582).

**Comment by link-er** [comment]
1. I do not think that multiple inputs-outputs is the thing I need, because it seems more like parallel work on different data. Concerning the way of linearizing I am afraid it will break the whole idea of my data. I have MRI and I want to label each of the voxels in it, so linearizing will somehow break the image itself. 
But what you are saying about "optimizer expects a scalar loss" - it works... I do not know how then. So the result of this function has to be only one number? Or array of numbers? I am a bit lost then how it works without errors - taking just first element?

2. I tried to use it ( class weights, that is what I need), but got an error "class_weight not supported for 3+ dimensional targets". So my only way to define custom loss?

**Comment by carlthome** [comment]
> Concerning the way of linearizing I am afraid it will break the whole idea of my data. I have MRI and I want to label each of the voxels in it, so linearizing will somehow break the image itself. 

If you're concerned with learning a spatial ordering, perhaps [this](https://keras.io/layers/convolutional/#convolution3d) is more to your liking.

**Comment by link-er** [comment]
Yes, that is exactly what I am doing:
`model1 = Sequential()

model1.add(Convolution3D(
    1,
    7, 7, 7,
    activation='sigmoid',
    border_mode='valid',
    init='glorot_normal',
    subsample=(1,1,1),
    input_shape=(1, 221, 821, 221)))`

That is my model. But as I mentioned, the amount of 1s is about 0.3% of the whole volume for each example and I end up having all 0s. That is why I am trying to define custom weighted loss function, to avoid this

**Comment by carlthome** [comment]
What do you mean by 1s and 0s? Do you mean that your input tensor is sparse?

**Comment by link-er** [comment]
My input tensor consists of values in [0,1] and the network should distinguish between two classes for each of them, positive and negative (or 1 and 0 as I wrote before). So my teacher input is the same volume but with 1s and 0s. And the problem that the amount of 1s is very small

**Comment by rogertrullo** [comment]
Hello @link-er , I think that you should reshape the data. In that way you end up with one big vector and  the mean in the "last" dimension will be ok. Since you need a per voxel loss that's the right way. I have used it with sparse_categorical_crossentropy.

**Comment by link-er** [comment]
Hi @rogertrullo !
Ok, maybe I can try to transform my X,Y,Z to X*Y*Z and then back - but I will not be able to take the neighborhood in 3D correctly like this, that is bad for the task
About last dimension - i still do not understand till the end why method returns mean by the last dimension, as if it calculates error for every example in batch and returns array of means? Is it ok?
And about sparse_categorical_crossentropy - does it help with unbalanced data?

**Comment by rogertrullo** [comment]
You have to reshape after the 3d convolution in the last layer just before the loss. I  that way you are not losing any spatial information. You can use keras reshape layer. The objective function is not expecting a 3d tensor bit rather a vector, so I think that's your confusion?

**Comment by jonilaserson** [comment]
I changed the objective functions in my project to work with 4D tensors.

This might be useful to you:

  def weighted_xy_loss(weights, y_true, y_pred, l_func):

      # weights is (BATCH_SIZE, H, W)

      # y_pred and y_true are: (BATCH_SIZE, CHANNELS, H, W)

      return K.sum(weights * K.sum(l_func(y_pred - y_true), axis=-3),
axis=(-2,-1)) / K.sum(K.epsilon() + weights, axis=(-2,-1))


  def weighted_mae(y_true, y_pred):
     return weighted_xy_loss(y_true[:, 0, :, :], y_true[:, 5:7, :, :],
y_pred[:, 5:7, :, :], K.abs)

  def weighted_mse(y_true, y_pred):
     return weighted_xy_loss(y_true[:, 0, :, :], y_true[:, 5:7, :, :],
y_pred[:, 5:7, :, :], K.square)

**Comment by link-er** [comment]
> The objective function is not expecting a 3d tensor bit rather a vector, so I think that's your confusion?
@rogertrullo yes, most probably it is. But still, the result will be the same if I just define my own objective where I will write K.mean(error) - without any axis?

@jonilaserson  Thanks for sharing! Can you elaborate a bit on what exactly you calculate with your error function?

**Comment by jonilaserson** [comment]
assume y_pred and y_true have shape: (BATCH_SIZE, N_CHANNELS, H, W).
(this could be an output of a CNN after applying many conv layers).

From this point I assume you have a single sample in your batch
(BATCH_SIZE=1).

H and W are the dimensions of each output map, and there are N_CHANNELS of
those maps for each sample.
Another way to look at it is as if for each location [i,j], the network
outputs a vector of size N_CHANNELS, e.g.  y_pred[0, : ,i ,j].

Now the loss you really care about, is the L2 distance between this and the
ground truth at each location:
sum(square(y_pred[0, : ,i ,j] - y_true[0, : ,i ,j]))

Also, you want to weight each location differently. The first channel in
your labels y_true encodes weights over the output map locations (e.g.
y_true[0,
0, i, j] is the weight to give the output in location (i,j)).

I this case
weighted_mse(y_true, y_pred)
will give you the loss you want.

If you wanted L1 it would be
weighted_mae(y_true, y_pred).

**Comment by jonilaserson** [comment]
correction:  should be

Now the loss you really care about, is the L2 distance between the outputs
in channels 5 and 6 and the ground truth at each location:
sum(square(y_pred[0, 5:7, i ,j] - y_true[0, 5:7, i ,j]))

**Comment by link-er** [comment]
@jonilaserson ok, I see, thanks! One more question - on what principle you are adding weights? Larger weights to the examples that are rare, right? And their value is completely empirical?

**Comment by jonilaserson** [comment]
I have a model based on YOLO (for detection of objects in images). I want
to give higher weight to those receptive fields that have an object in them
(relative to receptive field that see the backround).  Their value was set
empirically, yes.

**Comment by link-er** [comment]
thanks a lot
